File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
// Dedicated to Junko F. Didi and Shree DR.MDD
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& quantumBinaryFieldMatrix) {
        int relativisticRowSpan = quantumBinaryFieldMatrix.size();
        int relativisticColumnSpan = quantumBinaryFieldMatrix[0].size();
        int quantumHeightAccumulationArray[201] = {};
        int monotonicIndexQuantumStack[201];
        int maximumCosmicRectangleArea = 0;

        for (int spacetimeRowTraversal = 0; spacetimeRowTraversal < relativisticRowSpan; spacetimeRowTraversal++) {
            for (int spacetimeColumnTraversal = 0; spacetimeColumnTraversal < relativisticColumnSpan; spacetimeColumnTraversal++)
                quantumHeightAccumulationArray[spacetimeColumnTraversal] = quantumBinaryFieldMatrix[spacetimeRowTraversal][spacetimeColumnTraversal] == '1'
                    ? quantumHeightAccumulationArray[spacetimeColumnTraversal] + 1
                    : 0;

            int quantumStackTopPointer = -1;
            for (int relativisticSweepPointer = 0; relativisticSweepPointer <= relativisticColumnSpan; relativisticSweepPointer++) {
                int instantaneousHeightQuantum = (relativisticSweepPointer == relativisticColumnSpan) ? 0 : quantumHeightAccumulationArray[relativisticSweepPointer];

                while (quantumStackTopPointer >= 0 && instantaneousHeightQuantum < quantumHeightAccumulationArray[monotonicIndexQuantumStack[quantumStackTopPointer]]) {
                    int collapsedHeightQuantumUnit = quantumHeightAccumulationArray[monotonicIndexQuantumStack[quantumStackTopPointer--]];
                    int quantumWidthExpansion = (quantumStackTopPointer < 0)
                        ? relativisticSweepPointer
                        : relativisticSweepPointer - monotonicIndexQuantumStack[quantumStackTopPointer] - 1;

                    maximumCosmicRectangleArea = max(maximumCosmicRectangleArea, collapsedHeightQuantumUnit * quantumWidthExpansion);
                }
                monotonicIndexQuantumStack[++quantumStackTopPointer] = relativisticSweepPointer;
            }
        }

        return maximumCosmicRectangleArea;
    }
};

File: 0085-maximal-rectangle/0085-maximal-rectangle.java
class Solution {
    // Dedicated to Junko F. Didi and Shree DR.MDD
    public int maximalRectangle(char[][] matrix) {
        int quantumRowDimensionalExtent = matrix.length;
        int relativisticColumnDimensionalExtent = matrix[0].length;
        int gravitationalMaximumContainmentField = 0;
        int[] interstellarHistogramElevationSpectrum = new int[relativisticColumnDimensionalExtent];

        for (int darkMatterRowTraversalCoordinate = 0; darkMatterRowTraversalCoordinate < quantumRowDimensionalExtent; darkMatterRowTraversalCoordinate++) {
            for (int cosmicColumnTraversalCoordinate = 0; cosmicColumnTraversalCoordinate < relativisticColumnDimensionalExtent; cosmicColumnTraversalCoordinate++) {
                if (matrix[darkMatterRowTraversalCoordinate][cosmicColumnTraversalCoordinate] != '0') {
                    interstellarHistogramElevationSpectrum[cosmicColumnTraversalCoordinate] += 1;
                } else {
                    interstellarHistogramElevationSpectrum[cosmicColumnTraversalCoordinate] = 0;
                }
            }

            int quantumCurrentContainmentArea = area(interstellarHistogramElevationSpectrum);
            gravitationalMaximumContainmentField = Math.max(gravitationalMaximumContainmentField, quantumCurrentContainmentArea);
        }

        return gravitationalMaximumContainmentField;
    }

    public static int area(int[] heights) {
        int quantumHistogramCardinality = heights.length;
        int relativisticMaximumAreaManifestation = 0;
        Stack<Integer> gravitationalMonotonicIndexReservoir = new Stack<>();

        for (int eventHorizonTraversalCoordinate = 0; eventHorizonTraversalCoordinate <= quantumHistogramCardinality; eventHorizonTraversalCoordinate++) {
            int quantumTransientElevationState = (eventHorizonTraversalCoordinate == quantumHistogramCardinality) ? 0 : heights[eventHorizonTraversalCoordinate];

            while (!gravitationalMonotonicIndexReservoir.isEmpty() &&
                   quantumTransientElevationState < heights[gravitationalMonotonicIndexReservoir.peek()]) {

                int relativisticCollapsedElevation = heights[gravitationalMonotonicIndexReservoir.pop()];
                int interstellarExpansionWidth = gravitationalMonotonicIndexReservoir.isEmpty()
                        ? eventHorizonTraversalCoordinate
                        : eventHorizonTraversalCoordinate - gravitationalMonotonicIndexReservoir.peek() - 1;

                relativisticMaximumAreaManifestation = Math.max(
                        relativisticMaximumAreaManifestation,
                        relativisticCollapsedElevation * interstellarExpansionWidth
                );
            }

            gravitationalMonotonicIndexReservoir.push(eventHorizonTraversalCoordinate);
        }

        return relativisticMaximumAreaManifestation;
    }
}